### ISSUE TITLE
Upgrade lodash dependency to 4.*

### ISSUE THREAD

**Issue by dschissler** [AmpersandJS/ampersand-state#230]
`lodash` is currently on version `4.2.1` but all of `ampersand-*` packages are using `3`.  I think that a complete ampersand project wide upgrade should occur with perhaps a major version bump for each ampersand package.  lodash `4.0.0` has been out since `2016-01-12`.

Maybe wait another month and then bump it?

**Comment by wraithgar** [comment]
This is certainly something that would be nice to have, it's quite a large undertaking though.  Wondering if anyone feels they have the time to carve out to do this?

**Comment by cdaringe** [comment]
it may be easy-ish to do using [ampersand-family-meeting](https://github.com/cdaringe/ampersand-family-meeting/) .  could be a good first run with it.  need to finish the PR logic, which won't be hard.

**Comment by ianwremmel** [comment]
Would love to see this. I'm working on migrating a library to lodash 4 using babel-plugin-lodash and the library doubled in size because ampersand is pulling in lodash 3

**Comment by dschissler** [comment]
@cdaringe Is it really so difficult to update all packages to use a new lodash?

**Comment by cdaringe** [comment]
@dschissler, full time job, school, and life!  you're welcome to start staging these PRs on your own if you want.  we'd happily accept them :).  the upgrade is not so seamless

- not all modules are compatible with the upgrade immediately.  i did upgrade &-view, and it didn't pass tests. debugging required
- all lodash refs need to be converted to `lodash/mod` vs `lodash.mod` for package size reduction
- to be most effective, we'd want and try and hit ampersandjs-core modules (most important) but propose the updates to community modules as well.  with everyone aligned, bundle sizes continue to drop.  `npm search` will show you that the following pkgs exist:

1. ampersand-app
1. ampersand-array-checkbox-view
1. ampersand-array-input-view
1. ampersand-auto-closing-view
1. ampersand-autocomplete-view
1. ampersand-avatar-field
1. ampersand-avatar-input-view
1. ampersand-bootstrap-tagsinput-view
1. ampersand-bubble-prop
1. ampersand-calendar
1. ampersand-chart
1. ampersand-checkbox-view
1. ampersand-chess-state
1. ampersand-chess-view
1. ampersand-class-extend
1. ampersand-collection
1. ampersand-collection-filterable
1. ampersand-collection-fluxible-mixin
1. ampersand-collection-history
1. ampersand-collection-jquery-datatable
1. ampersand-collection-lodash-mixin
1. ampersand-collection-pouchdb-mixin
1. ampersand-collection-pusher-mixin
1. ampersand-collection-rest-mixin
1. ampersand-collection-sort-mixin
1. ampersand-collection-underscore-mixin
1. ampersand-collection-view
1. ampersand-controller-router
1. ampersand-date-view
1. ampersand-date-view2
1. ampersand-dependency-mixin
1. ampersand-dom
1. ampersand-dom-bindings
1. ampersand-domthing-mixin
1. ampersand-drawer-view
1. ampersand-events
1. ampersand-expanding-textarea-view
1. ampersand-express-router
1. ampersand-fetch
1. ampersand-file-drop-view
1. ampersand-filereader-input-view
1. ampersand-filter-tracker
1. ampersand-filtered-collection
1. ampersand-filtered-subcollection
1. ampersand-floatinglabel-input-view
1. ampersand-form-manager-view
1. ampersand-form-view
1. ampersand-fullcalendar-view
1. ampersand-grid-view
1. ampersand-grouped-collection-view
1. ampersand-hoodie-collection
1. ampersand-hoodie-mixin
1. ampersand-hoodie-model
1. ampersand-infinite-scroll
1. ampersand-input-view
1. ampersand-io
1. ampersand-io-collection
1. ampersand-io-model
1. ampersand-jid-datatype-mixin
1. ampersand-json-patch
1. ampersand-jsonapi-ajaxconfig
1. ampersand-jsonapi-collection
1. ampersand-jsonapi-model
1. ampersand-local-cache-mixin
1. ampersand-main-view
1. ampersand-mixins
1. ampersand-modal-view
1. ampersand-model
1. ampersand-model-optimistic-update-mixin
1. ampersand-model-patch-mixin
1. ampersand-model-pouchdb-mixin
1. ampersand-model-pusher-mixin
1. ampersand-model-validations-mixin
1. ampersand-movingobj
1. ampersand-multifield-view
1. ampersand-multiselect-view
1. ampersand-offline-first
1. ampersand-optimistic-sync
1. ampersand-paginated-subcollection
1. ampersand-pagination-mixin
1. ampersand-pikaday-view
1. ampersand-pouchdb-sync
1. ampersand-questionnaire-view
1. ampersand-radio-view
1. ampersand-react-connector
1. ampersand-react-mixin
1. ampersand-react-view
1. ampersand-react-view-mixin
1. ampersand-registry
1. ampersand-rest-collection
1. ampersand-router
1. ampersand-router-query-parameters
1. ampersand-sanitized-html-data-type
1. ampersand-search-select
1. ampersand-select-view
1. ampersand-single-sync
1. ampersand-state
1. ampersand-state-fluxible-mixin
1. ampersand-state-mixin-datatype-function
1. ampersand-state-mixin-datatype-iso-date
1. ampersand-store
1. ampersand-subcollection
1. ampersand-sync
1. ampersand-sync-adapter
1. ampersand-sync-localforage
1. ampersand-sync-promise
1. ampersand-sync-with-promise
1. ampersand-tab-view
1. ampersand-time-range
1. ampersand-version
1. ampersand-view
1. ampersand-view-conventions
1. ampersand-view-jquery-events
1. ampersand-view-jquery-mixin
1. ampersand-view-switcher
1. ampersand-virtual-dom-mixin
1. ampersand-webcam-snapshot-view
1. ampersand-websocket-hint
1. ampersand-webworker-demo
1. ampersand-wildemitter-datatype-mixin

so, in summary, (not to be curt), it is difficult and very time consuming.  that's why i started working on ampersand-family-agenda--to help automate these upgrades and to ease the synchronization of our loosely-coupled framework!

**Comment by pgilad** [comment]
Hey @cdaringe Thanks for all the details. Why is `lodash/get` better for package size reduction over `lodash.get` ? Doesn't the roll-up technique work the same on both?

**Comment by cdaringe** [comment]
hey @pgilad! how's it goin man?

j dalton explained in a presentation that the `lodash.xzy` **inlines the module _with_ all of its dependencies**.  so, if `.get` and `.set` both have a common dep (which they do, `_isKey`), you will consequently "over-bundle" it using the dot strategy.

by example:

```js
// get-set-dot-strategy.js
var set = require('lodash.set');
var get = require('lodash.get');
```
```js
// get-set-slash-strategy.js
var set = require('lodash/set');
var get = require('lodash/get');
```
```
# post browserify
cdieringer  staff  17038 Mar 25 00:34 get-set-dot-strategy.bundled.js
cdieringer  staff  15318 Mar 25 00:35 get-set-slash-strategy.bundled.js # less, yahtzee!
```

**Comment by dschissler** [comment]
@cdaringe Once everything is upgraded when using NPM 3 it would be the same since it installs flat dependencies.  Just sayin'.

**Comment by dschissler** [comment]
@cdaringe Would we bump the major version on every single post 1.0 module?  Also I think that if someone committed to upgrading about 20 core packages that they should be allowed a commit freeze for a defined period of time to get it all done.  It could become very annoying if tiny commits were sneaked into various packages during this conversion.

**Comment by dschissler** [comment]
I'm not sure how I would test it since I'm using only the following in my project:

```
ampersand-class-extend
ampersand-collection
ampersand-collection-view
ampersand-dom
ampersand-dom-bindings
ampersand-events
ampersand-router
ampersand-state
ampersand-version
ampersand-view
ampersand-view-switcher
```

**Comment by pgilad** [comment]
@cdaringe thanks for the info man!

I could get the upgrades in some of the packages if you want

**Comment by cdaringe** [comment]
> @cdaringe Once everything is upgraded when using NPM 3 it would be the same since it installs flat dependencies. Just sayin'.

hey @dschissler, the dot strategy _inlines the dependencies_, so even if npm3 flattens deps, they will still be duplicated in each lodash module :)

> I could get the upgrades in some of the packages if you want

@pgilad, if you'd like to! it's really about keepin' the users happy by this point, and churning to keep our deps up to date.  maybe greenkeeper is something we should consider?  i'm definitely not opposed to getting the ball rolling, i'm just swamped ATM

**Comment by wraithgar** [comment]
Given our test coverage, and the fact that greenkeeper makes PRs from local branches (i.e. the saucelabs credentials would show up) I'd be all for enabling greenkeeper.

**Comment by pgilad** [comment]
@cdaringe Can you verify this is the proper direction?
https://github.com/AmpersandJS/ampersand-events/pull/15

**Comment by wraithgar** [comment]
Looks good @pgilad and that's a good module to start in cause it's at the top of the dependency tree.
